src/lib/pubkey/curve448/curve448_utils/curve448_gf.cpp

```diff
@@ -0,0 +1,399 @@
+   const uint64_t z = diff - static_cast<uint64_t>(*borrow);
+   *borrow = static_cast<uint64_t>(borrow_a_min_b) | static_cast<uint64_t>(z > diff);
+   return z;
+}
+
+/**
+ * @brief Reduce the result of a addition modulo 2^448 - 2^224 - 1.
+ *
+ * Algorithm 1 of paper "Reduction Modulo 2^448 - 2^224 - 1", from line 27.
+ *
+ * @param h_3 Output
+ * @param h_1 Input
+ */
+void reduce_after_add(std::span<uint64_t, 7> h_3, std::span<const uint64_t, 8> h_1) {
+   std::array<uint64_t, 8> h_2;
+   bool carry;
+
+   // Line 27+ (of the paper's algorithm 1)
+   h_2[0] = u64_add(h_1[0], h_1[7], &carry);
+
+   h_2[1] = u64_add(h_1[1], carry, &carry);
+   h_2[2] = u64_add(h_1[2], carry, &carry);
+
+   // Line 30
+   h_2[3] = u64_add_with_carry(h_1[3], h_1[7] << 32, &carry);
+
+   // Line 31+
+   h_2[4] = u64_add(h_1[4], carry, &carry);
+   h_2[5] = u64_add(h_1[5], carry, &carry);
+   h_2[6] = u64_add(h_1[6], carry, &carry);
+
+   h_2[7] = carry;
+
+   h_3[0] = u64_add(h_2[0], h_2[7], &carry);
+   h_3[1] = u64_add(h_2[1], carry, &carry);
+   h_3[2] = u64_add(h_2[2], carry, &carry);
+   // Line 37
+   h_3[3] = h_2[3] + (h_2[7] << 32) + carry;
+
+   // Line 38
+   h_3[4] = h_2[4];
+   h_3[5] = h_2[5];
+   h_3[6] = h_2[6];
+}
+
+/**
+ * @brief Reduce the result of a addition modulo 2^448 - 2^224 - 1.
+ *
+ * Algorithm 1 of paper "Reduction Modulo 2^448 - 2^224 - 1".
+ */
+void reduce_after_mul(std::span<uint64_t, 7> out, std::span<const uint64_t, 14> in) {
+   std::array<uint64_t, 8> r;
+   std::array<uint64_t, 8> s;
+   std::array<uint64_t, 8> t_0;
+   std::array<uint64_t, 8> h_1;
+
+   bool carry;
+
+   // Line 4 (of the paper's algorithm 1)
+   r[0] = u64_add(in[0], in[7], &carry);
+
+   // Line 5-7
+   for(size_t i = 1; i < 7; ++i) {
+      r[i] = u64_add_with_carry(in[i], in[i + 7], &carry);
+   }
+   r[7] = carry;
+   s[0] = r[0];
+   s[1] = r[1];
+   s[2] = r[2];
+   // Line 10
+   s[3] = u64_add(r[3], in[10] & 0xFFFFFFFF00000000, &carry);
+   // Line 11-13
+   for(size_t i = 4; i < 7; ++i) {
+      s[i] = u64_add_with_carry(r[i], in[i + 7], &carry);
+   }
+   s[7] = r[7] + carry;
+
+   // Line 15-17
+   for(size_t i = 0; i < 3; ++i) {
+      t_0[i] = (in[i + 11] << 32) | (in[i + 10] >> 32);
+   }
+   // Line 18
+   t_0[3] = (in[7] << 32) | (in[13] >> 32);
+   // Line 19-21
+   for(size_t i = 4; i < 7; ++i) {
+      t_0[i] = (in[i + 4] << 32) | (in[i + 3] >> 32);
+   }
+   h_1[0] = u64_add(s[0], t_0[0], &carry);
+   // Line 23-25
+   for(size_t i = 1; i < 7; ++i) {
+      h_1[i] = u64_add_with_carry(s[i], t_0[i], &carry);
+   }
+   h_1[7] = s[7] + carry;
+
+   reduce_after_add(out, h_1);
+}
+
+constexpr size_t words_per_uint64 = 8 / sizeof(word);
+
+void gf_mul(std::span<uint64_t, 7> out, std::span<const uint64_t, 7> a, std::span<const uint64_t, 7> b) {
+   std::array<uint64_t, 14> ws;
+   if constexpr(std::same_as<uint64_t, word>) {
+      // Reinterpret cast to itself to prevent compiler errors on non 64-bit systems
```

Comment:
Yeah, I'm also not happy with this reinterpret_cast workaround. But I also decided #ifdef is worse.

Reply:
I would myself lean to the preprocessor here. But either way is ugly so I am not too concerned either way.

I think the real fix is to have all of the `mp` algorithms templatized over the word type so that you can just use `bigint_mul_7` on `uint64_t` regardless of word size. This relates to some work I’ve mentioned for ellptic curve refactorings that I may or may not have time to pursue this year.